pallets/poa/src/lib.rs

```diff
@@ -0,0 +1,710 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Decode;
+/// Pallet to add and remove validators.
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch, ensure, fail,
+    sp_runtime::{print, SaturatedConversion},
+    traits::{Currency, Get, Imbalance, OnUnbalanced},
+};
+use frame_system::{self as system, ensure_root};
+use log::{debug, warn};
+use sp_std::prelude::Vec;
+
+extern crate alloc;
+use alloc::collections::BTreeSet;
+
+#[cfg(test)]
+mod tests;
+
+/// The pallet's configuration trait.
+pub trait Trait: system::Trait + pallet_session::Trait + pallet_authorship::Trait {
+    /// The overarching event type.
+    type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+
+    /// Epoch length in number of slots
+    type MinEpochLength: Get<u64>;
+
+    /// Maximum no. of active validators allowed
+    type MaxActiveValidators: Get<u8>;
+
+    type Currency: Currency<Self::AccountId>;
+}
+
+// This pallet's storage items.
+decl_storage! {
+
+    trait Store for Module<T: Trait> as PoAModule {
+        /// List of active validators. Maximum allowed are `MaxActiveValidators`
+        ActiveValidators get(fn active_validators) config(): Vec<T::AccountId>;
```

Comment:
> This is a mirror of the public pallet_session::Validators.

We talked about this in the call last week. Is there something else you are pointing out.

> Some logic in this module assumes that poa::ActiveValidators is equal to pallet_session::Validators.

This module fess the validator set (`ActiveValidators`) to `pallet_session`.

> This assumption is likely to be sometimes untrue.

When?

Reply:
> When

After `should_end_session()` is called and before `pallet_session` updates the actual validator set.

Reply:
Because that change is letting the session module to pick up new validators

Reply:
I still think it's a problem.

Reply:
> implicit invariant (poa::ActiveValidators::get() == pallet_session::Validators::get()) must be maintained

Why? `ActiveValidators` is used to set `pallet_session::Validators`. What is issue when the invariant is false?

> The scope of the invariant (the code for which a contributor must be aware of the invariant when making changes) is large. The scope is the entire module which is large and growing.

The contributor does not need to read or write `pallet_session::Validators`. 

> There are not checks to validate the invariant. It may be invalid now and future changes (like updates to substrate) may make it invalid without us knowing.

What do you suggest testing? Does testing call to `new_session` returning same result as `ActiveValidators` (depending on the flag) needed.

> the invariant is not documented

Will document.

> I expect the lag between poa::ActiveValidators and pallet_session::Validators to increase the likelyhood of undetected bugs in validation logic or in validator remittance calculations.

The difference `ActiveValidators` between `pallet_session::Validators` only exists when epoch is being terminated (except genesis) and changes from `ActiveValidators` to `pallet_session::Validators` are propagated in the same thread. If you don't see an issue now, lets not block.